### PR TITLE
Allow retries for statuses other than 429 in streaming bulk

### PR DESCRIPTION
**Pull request by ayayron**
Closes https://github.com/elastic/elasticsearch-py/issues/1004.

This updates https://github.com/elastic/elasticsearch-py/pull/1005 to work for both the async and sync client as well as adding tests.

### CODE REVIEW

**Comment by pquentin** [comment]
buildkite test this please

File: elasticsearch/_async/helpers.py [hunk 1]
```diff
@@ -184,10 +185,11 @@ async def async_streaming_bulk(
     entire input is consumed and sent.
 
     If you specify ``max_retries`` it will also retry any documents that were
-    rejected with a ``429`` status code. To do this it will wait (**by calling
-    asyncio.sleep**) for ``initial_backoff`` seconds and then,
-    every subsequent rejection for the same chunk, for double the time every
-    time up to ``max_backoff`` seconds.
+    rejected with a ``429`` status code. Use ``retry_on_status`` to
+    configure which status codes will be retried. To do this it will wait
+    (**by calling time.sleep which will block**) for ``initial_backoff`` seconds
```
Comment (miguelgrinberg) [inline elasticsearch/_async/helpers.py:190]:
```suggestion
    (**by calling asyncio.sleep which will block**) for ``initial_backoff`` seconds
```

File: elasticsearch/_async/helpers.py [hunk 2]
```diff
@@ -264,11 +272,11 @@ async def map_actions() -> AsyncIterable[_TYPE_BULK_ACTION_HEADER_AND_BODY]:
                 ):
                     if not ok:
                         action, info = info.popitem()
-                        # retry if retries enabled, we get 429, and we are not
-                        # in the last attempt
+                        # retry if retries enabled, we are not in the last attempt,
+                        # and status not in retry_on_status (defaulting to 429)
```
Comment (miguelgrinberg) [inline elasticsearch/_async/helpers.py:276]:
```suggestion
                        # and status in retry_on_status (defaulting to 429)
```

File: elasticsearch/helpers/actions.py [hunk 3]
```diff
@@ -458,11 +466,11 @@ def streaming_bulk(
                     ):
                         if not ok:
                             action, info = info.popitem()
-                            # retry if retries enabled, we get 429, and we are not
-                            # in the last attempt
+                            # retry if retries enabled, we are not in the last attempt,
+                            # and status not in retry_on_status (defaulting to 429)
```
Comment (miguelgrinberg) [inline elasticsearch/helpers/actions.py:470]:
```suggestion
                            # and status in retry_on_status (defaulting to 429)
```

**Review by miguelgrinberg — APPROVED** [review]
LGTM, but some comments seemed inaccurate to me so I've added some suggested improvements.

**Comment by pquentin** [comment]
buildkite test this please

**Comment by pquentin** [comment]
Thank you for your contribution, it is now released as part of 8.16.0: https://github.com/elastic/elasticsearch-py/releases/tag/v8.16.0

**Pull request merged**